### PR TITLE
apply to acl commit

### PR DESCRIPTION
**Pull request by bansalpradeep**
Minitest result

TestInterface#test_interface_ipv4_address = 5.22 s = .
TestInterface#test_interface_speed_change = 4.99 s = .
TestInterface#test_interface_description_nil = 0.81 s = .
TestInterface#test_interface_create_valid = 0.82 s = .
TestInterface#test_interface_encapsulation_dot1q_change = 5.07 s = .
TestInterface#test_interface_speed_invalid = 1.16 s = .
TestInterface#test_negotiate_auto_ethernet = 3.57 s = S
TestInterface#test_interface_duplex_invalid = 4.18 s = .
TestInterface#test_negotiate_auto_portchannel = 6.90 s = .

Finished in 410.424292s, 0.1096 runs/s, 0.7967 assertions/s.

  1) Skipped:
TestInterface#test_negotiate_auto_ethernet [tests/test_interface.rb:717]:
Skip test: Interface type does not allow config change

45 runs, 327 assertions, 0 failures, 0 errors, 1 skips
Coverage report generated for MiniTest to /home/puppet-agent/pradeepb/prad_latest/cisco-network-node-utils/coverage. 628 / 811 LOC (77.44%) covered.

### CODE REVIEW

**Comment by chrisvanheuveln** [comment]
1. Please run rubocop before submitting a PR.
2. You also need to update CHANGELOG.md

https://github.com/cisco/cisco-network-node-utils/blob/develop/CONTRIBUTING.md#making-changes

File: lib/cisco_node_utils/cmd_ref/interface.yaml [hunk 1]
```diff
@@ -12,6 +12,13 @@ access_vlan:
   kind: int
   default_value: 1
 
+acl_ipv4:
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/cmd_ref/interface.yaml]: Please rename these to have the AFI as the prefix. Also, this needs to have four properties:

`ipv4_acl_in`
`ipv4_acl_out`
`ipv6_acl_in`
`ipv6_acl_out`

e.g.:

```
ipv4_acl_in:
  config_get_token_append: '/^ip access-group (\S+) in/'
  config_set_append: "%s ip access-group %s in"
  default_value: ~
```

File: lib/cisco_node_utils/interface.rb [hunk 2]
```diff
@@ -74,6 +74,20 @@ def access_vlan=(vlan)
       raise "[#{@name}] '#{e.command}' : #{e.clierror}"
     end
 
+    def acl
+      config_get('interface', 'acl', @name)
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/interface.rb]: How could this possibly work? You don't have an `acl` tag in the yaml file?

There are fundamental problems here. The code review process is not really the place for figuring out how your code is supposed to work. Please work with Saqib and Shengjun to get this code straightened out. I'll make a few comments here but it needs a total rewrite. 

If you look at the surrounding code for this provider there are plenty of examples for how to add a new property.

Your getters should follow this convention:

```
    def ipv4_acl_in
      config_get('interface', 'ipv4_acl_in', @name)
    end
```

Your setters should look something like this:

```
    def ipv4_acl_in=(val)
      if val
        state = ''
      else
        state = 'no'
        val = ipv4_acl_in
      end
      config_set('interface', 'ipv4_acl_in', @name, state, val)
    end
```

Your default getters should be:

```
    def default_ipv4_acl_in
      config_get_default('interface', 'default_ipv4_acl_in', @name)
    end
```

File: lib/cisco_node_utils/interface.rb [hunk 3]
```diff
@@ -202,7 +216,8 @@ def fex_feature_set(fex_set)
         config_set('fex', 'feature', '')
       when :disabled
         config_set('fex', 'feature', 'no') if curr == :enabled
-        return
+        return/
+
```
Comment (chrisvanheuveln) [inline lib/cisco_node_utils/interface.rb]: Please remove these two changes.

File: tests/test_interface.rb [hunk 4]
```diff
@@ -361,6 +361,7 @@ def validate_vrf(inttype_h)
     end
   end
 
+  # def validate_acl(inttype_h)
```
Comment (chrisvanheuveln) [inline tests/test_interface.rb]: remove

File: tests/test_interface.rb [hunk 5]
```diff
@@ -778,6 +779,29 @@ def test_interface_ipv4_addr_mask_set_netmask_invalid
     interface_ethernet_default(interfaces_id[0])
   end
 
+  def test_create_acl
```
Comment (chrisvanheuveln) [inline tests/test_interface.rb]: You'll need to rewrite these tests. Here's a simple example, just repeat it for ipv6:

```
  def test_ipv4_acl
    # Sample cli:
    #
    #   interface Ethernet1/1
    #     ip access-group foo in
    #     ip access-group bar out
    #
    interface_ethernet_default(interfaces[0])
    i = Interface.new(interfaces[0])

    i.ipv4_acl_in = 'foo'
    assert_equal('foo', i.ipv4_acl_in)
    i.ipv4_acl_out = 'bar'
    assert_equal('bar', i.ipv4_acl_out)

    i.ipv4_acl_in = 'foo2'
    assert_equal('foo2', i.ipv4_acl_in)
    i.ipv4_acl_out = 'bar2'
    assert_equal('bar2', i.ipv4_acl_out)

    i.ipv4_acl_in = i.default_ipv4_acl_in
    assert_equal('foo2', i.ipv4_acl_in)
    i.ipv4_acl_out = i.default_ipv4_acl_out
    assert_equal('bar2', i.ipv4_acl_out)
  end
```

File: tests/test_interface.rb [hunk 6]
```diff
@@ -1030,6 +1054,7 @@ def test_interface_ipv4_all_interfaces
     # pre-configure
     inttype_h = config_from_hash(inttype_h)
 
+    print inttype_h
```
Comment (chrisvanheuveln) [inline tests/test_interface.rb]: remove

File: tests/test_interface.rb [hunk 7]
```diff
@@ -1048,6 +1073,7 @@ def test_interface_ipv4_all_interfaces
       validate_ipv4_redirects(inttype_h)
       validate_interface_shutdown(inttype_h)
       validate_vrf(inttype_h)
+      # validate_acl(inttype_h)
```
Comment (chrisvanheuveln) [inline tests/test_interface.rb]: remove

**Comment by chrisvanheuveln** [comment]
_"ACL has no default values."_  Not exactly true. The default state of `ip access-group <acl> in` is not configured (in the YAML file this is `default_value: ~`, which means return nil).

You need a default getter (as shown in my earlier comment) to be able to turn off an existing state; ie. your manifest wants to ensure that `ip access-group <acl> in` is not present on the interface. You also need default support to meet requirements for Chef, which sets all non-explicit properties to their default state.

File: CHANGELOG.md [hunk 8]
```diff
@@ -24,6 +24,7 @@ Changelog
 * Extend interface with `channel_group` attribute and support for fabric path attributes
 * Extend vrf with `vni` attribute
 * Extend vlan with `mode` attribute
+* Extend interface with access lists attributes (@bansalpradeep)
```
Comment (chrisvanheuveln) [inline CHANGELOG.md]: Just update the existing interface bullet:

```
-  * Extend interface with `channel_group` attribute and support for fabric path attributes
+  * Extend interface with attributes: `channel_group`, `ipv4_acl_in`, `ipv4_acl_out`, `ipv6_acl_in`, `ipv6_acl_out`. Also add support for fabric path attributes
```

File: tests/test_interface.rb [hunk 9]
```diff
@@ -778,6 +778,48 @@ def test_interface_ipv4_addr_mask_set_netmask_invalid
...
+
+    intf.ipv4_acl_in = 'v4acl1'
+    assert_equal('v4acl1', intf.ipv4_acl_in)
+    intf.ipv4_acl_out = 'v4acl2'
+    assert_equal('v4acl2', intf.ipv4_acl_out)
+
+    intf.ipv4_acl_in = 'v4acl3'
+    assert_equal('v4acl3', intf.ipv4_acl_in)
+    intf.ipv4_acl_out = 'v4acl4'
+    assert_equal('v4acl4', intf.ipv4_acl_out)
```
Comment (chrisvanheuveln) [inline tests/test_interface.rb:805]: Please add the default api tests as shown in my earlier comment.

File: tests/test_interface.rb [hunk 10]
```diff
@@ -778,6 +778,48 @@ def test_interface_ipv4_addr_mask_set_netmask_invalid
...
+    assert_equal('v6acl1', intf.ipv6_acl_in)
+    intf.ipv6_acl_out = 'v6acl2'
+    assert_equal('v6acl2', intf.ipv6_acl_out)
+
+    intf.ipv6_acl_in = 'v6acl3'
+    assert_equal('v6acl3', intf.ipv6_acl_in)
+    intf.ipv6_acl_out = 'v6acl4'
+    assert_equal('v6acl4', intf.ipv6_acl_out)
+  end
+
```
Comment (chrisvanheuveln) [inline tests/test_interface.rb:853]: Will these tests work if the acl does not currently exist? ie. If I try to run this test on my own testbed will it pass? Checking this on an n3k it appears that it won't.

Comment (bansalpradeep) [inline tests/test_interface.rb:853]: ACL needs to be created first. I had ACLs already on my switch. I have modified the tests to create ACLs also within the test itself

**Comment by chrisvanheuveln** [comment]
Manual commit & merge was required due to merge conflicts with CHANGELOG.md.

See merge:  9bf2c2032461a35cc2afc92b433b9d9f2d5bd3e1

Closing this PR because the tools cannot do it automatically.

**Pull request closed without merging**